proof_test.go

```diff
@@ -0,0 +1,120 @@
+package merkletree
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+	"github.com/coniks-sys/libmerkleprefixtree-go/internal"
+)
+
+func computeLeafHash(ap *AuthenticationPath) (leafHash []byte) {
+	leaf := ap.Leaf()
+	if leaf.Commitment() != nil {
+		// user leaf node
+		leafHash = crypto.Digest(
+			[]byte{LeafIdentifier},              // K_leaf
+			[]byte(ap.TreeNonce()),              // K_n
+			[]byte(ap.Index()),                  // i
+			[]byte(util.IntToBytes(ap.Level())), // l
+			[]byte(leaf.Commitment()),           // commit(key|| value)
+		)
+	} else {
+		// empty leaf node
+		leafHash = crypto.Digest(
+			[]byte{EmptyBranchIdentifier},       // K_empty
+			[]byte(ap.TreeNonce()),              // K_n
+			[]byte(ap.Index()),                  // i
+			[]byte(util.IntToBytes(ap.Level())), // l
+		)
+	}
+	return
+}
+
+func computeProofHash(ap *AuthenticationPath, depth int) (leftHash, rightHash []byte) {
```

Comment:
> the flow of this function is mostly the same as ``RecomputeHash()`` method of ``MerkleTree``.

I thought it would have been more intuitive to actually start the function at the deepest level since you are recomputing the hashes from the leaf tree up. It wasn't immediately clear that this is what ```computeProofHash()``` is doing (unlike in ```RecomputeHash()``` it's immediately clear how you're traversing the tree). But I like your implementation, too.